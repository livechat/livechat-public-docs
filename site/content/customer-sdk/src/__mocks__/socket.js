import mitt from '@livechat/mitt'
import fs from 'fs'
import path from 'path'
import cloneDeep from 'lodash/fp/cloneDeep'
import compose from 'lodash/fp/compose'
import find from 'lodash/fp/find'
import flatMap from 'lodash/fp/flatMap'
import get from 'lodash/fp/get'
import groupBy from 'lodash/fp/groupBy'
import keyBy from 'lodash/fp/keyBy'
import last from 'lodash/fp/last'
import map from 'lodash/fp/map'
import mapValues from 'lodash/fp/mapValues'
import merge from 'lodash/fp/merge'
import omit from 'lodash/fp/omit'
import pick from 'lodash/fp/pick'
import unionBy from 'lodash/fp/unionBy'

const unionById = unionBy(get('id'))

const FIXTURE_PATH = path.join(__dirname, '../../fixtures')
const fixturesCache = {}

const loadFixture = fixturePath => {
	if (fixturesCache[fixturePath]) {
		return fixturesCache[fixturePath]
	}
	return (fixturesCache[fixturePath] = fs.readFileSync(fixturePath, { encoding: 'utf8' }))
}

const TEST_CUSTOMER_ID = JSON.parse(loadFixture(path.join(FIXTURE_PATH, 'users/customer1.json'))).id

let serverState

const getChatsSummary = pagination => {
	return {
		chats_summary: map(chat => {
			return {
				...pick(['id', 'order', 'users'], chat),
				last_event_per_type: compose(
					mapValues(
						compose(event => {
							const { id: threadId, order: threadOrder } = find(
								thread => thread.events.some(({ id }) => event.id === id),
								chat.threads,
							)
							return {
								event: cloneDeep(event),
								thread_id: threadId,
								thread_order: threadOrder,
							}
						}, last),
					),
					groupBy(get('type')),
					flatMap(get('events')),
				)(chat.threads),
			}
		}, serverState.chats),
		total_chats: Object.keys(serverState.chats).length,
	}
}

// should consider 2 scenarios - with active and inactive thread
const getSentEvent = ({ chat_id: chat, event }) => {
	const lastThread = last(serverState.chats[chat].threads)

	const sentEvent = {
		type: event.type,
		id: `sent_event_${ ++serverState.sentEventsCounter }`,
		author_id: TEST_CUSTOMER_ID,
		order: lastThread.active ? last(lastThread.events).order + 1 : 1,
		timestamp: Math.floor(Date.now() / 1000),
	}

	if (typeof event.custom_id === 'string') {
		sentEvent.custom_id = event.custom_id
	}

	switch (event.type) {
	case 'message':
		if (typeof event.text !== 'string') {
			throw new Error(
				JSON.stringify({
					message: '`event.text` is required',
					type: 'validation',
				}),
			)
		}
		sentEvent.text = event.text
		break
	default:
		Object.assign(sentEvent, omit(['type', 'custom_id'], event))
		break
	}

	return {
		thread_id: lastThread.active ? lastThread.id : `created_thread_${ ++serverState.createdThreadsCounter }`,
		event: sentEvent,
	}
}

const createResponse = ({ request_id: id, action }, payload) => ({
	type: 'response',
	success: true,
	request_id: id,
	action,
	payload,
})

const createServerError = ({ request_id: id, action }, payload) => ({
	type: 'response',
	success: false,
	request_id: id,
	action,
	payload: { error: JSON.parse(payload) },
})

const createPushResponse = (request, response) => {
	switch (request.action) {
	case 'send_event':
		const chatId = request.payload.chat_id
		const serverChat = serverState.chats[chatId]
		const lastThread = last(serverChat.threads)

		if (lastThread.id !== response.payload.thread_id) {
			const createdThread = {
				active: true,
				order: lastThread.order + 1,
				id: response.payload.thread_id,
				users: [find(user => user.id === TEST_CUSTOMER_ID, serverChat.users)].map(cloneDeep),
				events: [cloneDeep(response.payload.event)],
			}
			serverState.chats[chatId].threads.push(createdThread)
			return {
				...response,
				type: 'push',
				action: 'incoming_chat_thread',
				payload: {
					chat: {
						id: chatId,
						users: serverChat.users.map(cloneDeep),
						thread: cloneDeep({
							...omit(['users'], createdThread),
							user_ids: createdThread.users.map(get('id')),
						}),
					},
				},
			}
		}

		lastThread.events.push(cloneDeep(response.payload.event))
		return merge(response, {
			type: 'push',
			action: 'incoming_event',
			payload: {
				chat_id: chatId,
			},
		})
	case 'update_customer':
		return merge(response, {
			type: 'push',
			action: 'customer_updated',
		})
	default:
		throw new Error('Unknown request.action')
	}
}

const handleRequest = (emitter, request) => {
	switch (request.action) {
	case 'get_chats_summary':
		Promise.resolve().then(() => {
			const response = createResponse(request, getChatsSummary(request.payload))
			emitter.emit('message', response)
		})
		return
	case 'login':
		Promise.resolve().then(() => {
			emitter.emit(
				'message',
				createResponse(request, {
					customer_id: TEST_CUSTOMER_ID,
					static_config_path: 'https://some.url.com',
				}),
			)
		})
		return
	case 'send_event':
		Promise.resolve()
			.then(() => {
				const response = createResponse(request, getSentEvent(request.payload))
				emitter.emit('message', response)
				return response
			})
			.then(response => {
				const pushResponse = createPushResponse(request, response)
				emitter.emit('message', pushResponse)
			})
			.catch(err => {
				emitter.emit('message', createServerError(request, err.message))
			})
		return
	case 'update_customer':
		Promise.resolve()
			.then(() => {
				// this flow need some serious work
				// 1. response data needs to be merged: sentData+serverData
				// 2. returned name should default to Customer
				// 3. payload shape need to be validated
				const response = createResponse(request, { customer: merge({}, request.payload) })
				emitter.emit('message', response)
				return response
			})
			.then(response => {
				const pushResponse = createPushResponse(request, response)
				emitter.emit('message', pushResponse)
			})
		return
	default:
		throw new Error('Unknown request.action')
	}
}

const emptyServerState = {
	chats: {},
	customer: TEST_CUSTOMER_ID,
	sentEventsCounter: 0,
	createdThreadsCounter: 0,
}

export const server = {
	addChats(fixtureNames) {
		const chats = fixtureNames.map(fixtureName => {
			const chat = JSON.parse(loadFixture(path.join(FIXTURE_PATH, 'chats', `${ fixtureName }.json`)))
			const lastThread = last(chat.threads)
			const threads = chat.threads.map(thread => {
				return {
					...thread,
					active: thread === lastThread ? thread.active === true : false,
					users: thread.users.map(user => {
						return JSON.parse(loadFixture(path.join(FIXTURE_PATH, 'users', `${ user.replace('fixture_', '') }.json`)))
					}),
				}
			})

			return {
				...chat,
				threads,
				users: threads.map(({ users }) => users).reduce((accUsers, users) => unionById(accUsers, users), []),
			}
		})
		serverState.chats = keyBy('id', chats)
	},
	clearHistory() {
		server.setHistory(emptyServerState)
	},
	setHistory(state) {
		serverState = cloneDeep(state)
	},
}

export default {
	create() {
		const emitter = mitt()
		let connected = false
		const connect = () => {
			Promise.resolve().then(() => {
				connected = true
				emitter.emit('connect')
			})
		}
		return {
			connect,
			disconnect: () => (connected = false),
			emit(request) {
				handleRequest(emitter, request)
			},
			on: emitter.on,
			off: emitter.off,
		}
	},
}
