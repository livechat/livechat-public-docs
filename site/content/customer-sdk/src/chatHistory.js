// @flow
import * as serverDataParser from './serverDataParser'
import { flatMap, splitRightWhenAccum } from '@livechat/data-utils'
import type { ChatId, CustomerSDK, Store, ThreadId, ThreadSummary } from './types'

type QueuedTask = {|
	resolve: any => void,
	reject: any => void,
|}

type HistoryState = {|
	done: boolean,
	idle: boolean,
	leftToFetch: number,
	queuedTasks: QueuedTask[],
	totalThreads: number,
	threadsSummary: ?(ThreadSummary[]),
	threadsToFetch: ThreadId[],
|}

const HISTORY_EVENT_COUNT_TARGET = 50

const splitSummary = threadsSummary => {
	const [leftover, targeted] = splitRightWhenAccum(
		({ totalEvents }, acc) => {
			const sum = acc + totalEvents
			return [sum >= HISTORY_EVENT_COUNT_TARGET, sum]
		},
		0,
		threadsSummary,
	)

	return [leftover, targeted.map(threadSummary => threadSummary.id)]
}

const distributeSummary = (state, threadsSummary) => {
	const result = splitSummary(threadsSummary)
	state.threadsSummary = result[0]
	state.threadsToFetch = result[1]
}

const createState = (): HistoryState => ({
	done: false,
	idle: true,
	leftToFetch: Infinity,
	queuedTasks: [],
	totalThreads: Infinity,
	threadsSummary: null,
	threadsToFetch: [],
})

const loadThreads = (api, chat, threadsToFetch) => {
	return api.getChatThreads(chat, threadsToFetch)
}

const loadLatest = (api, chat) =>
	api.getChatThreadsSummary(chat, { offset: 0 }).then(({ threadsSummary, totalThreads }) => {
		const [threadsSummaryLeftover, threadsToFetch] = splitSummary(threadsSummary)

		return Promise.all([
			{ threadsSummary: threadsSummaryLeftover, totalThreads },
			loadThreads(api, chat, threadsToFetch),
		])
	})

const loadNextSummaryBatch = (api, chat, state) =>
	api.getChatThreadsSummary(chat, { offset: 0 }).then(({ totalThreads }) => {
		const offset = totalThreads - state.leftToFetch
		return api.getChatThreadsSummary(chat, { offset })
	})

const onLoadedThreads = (store, state, threads) => {
	state.idle = true
	state.threadsToFetch = []
	state.leftToFetch -= threads.length

	const events = flatMap(thread => thread.events, threads)

	store.dispatch({
		type: 'emit_events',
		payload: threads.map(thread => ['thread_summary', serverDataParser.parseThreadSummary(thread)]),
	})

	if (state.leftToFetch === 0) {
		state.done = true
		return { value: events, done: true }
	}

	return { value: events, done: false }
}

const onFailedLoad = (state, err) => {
	state.idle = true
	state.queuedTasks.forEach(({ reject }) => reject(err))
	return err
}

export default (store: Store, api: CustomerSDK, chat: ChatId) => {
	const historyState = createState()

	const next = (resolve, reject) => {
		if (historyState.done) {
			resolve({ done: true })
			return
		}
		if (!historyState.idle) {
			historyState.queuedTasks.push({ resolve, reject })
			return
		}

		let task
		historyState.idle = false

		if (!historyState.threadsSummary) {
			task = loadLatest(api, chat).then(([{ threadsSummary, totalThreads }, threads]) => {
				historyState.threadsSummary = threadsSummary
				historyState.totalThreads = historyState.leftToFetch = totalThreads
				return threads
			})
		} else if (historyState.threadsToFetch.length > 0) {
			task = loadThreads(api, chat, historyState.threadsToFetch)
		} else if (historyState.threadsSummary.length > 0) {
			distributeSummary(historyState, historyState.threadsSummary)
			task = loadThreads(api, chat, historyState.threadsToFetch)
		} else {
			task = loadNextSummaryBatch(api, chat, historyState).then(({ threadsSummary }) => {
				distributeSummary(historyState, threadsSummary)
				return loadThreads(api, chat, historyState.threadsToFetch)
			})
		}

		task
			.then(threads => {
				resolve(onLoadedThreads(store, historyState, threads))

				const queuedTask = historyState.queuedTasks.shift()

				if (!queuedTask) {
					return
				}

				next(queuedTask.resolve, queuedTask.reject)
			})
			.catch(err => reject(onFailedLoad(historyState, err)))
	}

	return {
		next() {
			return new Promise(next)
		},
	}
}
