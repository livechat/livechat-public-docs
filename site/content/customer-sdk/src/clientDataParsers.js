// @flow
import { isObject, pickOwn } from '@livechat/data-utils'
import type { ChatId, EventOptionalProps, RNFile } from './types'

export const parseRatingProperties = event => ({ rating: pickOwn(['score', 'comment'], event) })

// TODO: we could validate and throw here
// but should we? maybe only in DEV mode?
export const parseEvent = event => {
	let parsed = {}
	parsed.type = event.type
	switch (event.type) {
	case 'message':
		parsed.text = event.text
		break
	case 'rating':
		parsed.type = 'annotation'
		parsed.annotation_type = 'rating'
		parsed.properties = parseRatingProperties(event)
		break
	default:
		parsed = event
		break
	}

	if (typeof event.customId === 'string') {
		parsed.custom_id = event.customId
	}

	if (event.postback) {
		parsed.postback = event.postback
	}

	if (isObject(event.properties)) {
		parsed.properties = event.properties
	}

	return parsed
}

export const parseFileData = (
	chat: ChatId,
	data: {
		...EventOptionalProps,
		file: Blob | RNFile,
	},
) => {
	const payload = {}
	payload['payload.chat_id'] = chat
	payload['payload.file'] = data.file
	if (data.customId) {
		payload['payload.custom_id'] = data.customId
	}
	return payload
}

export const parseStartChatData = ({ scope = { type: 'license' }, events, properties }) => {
	const data = {}
	data.chat = {}
	data.routing_scope = scope

	if (events !== undefined) {
		data.chat.thread = { events: events.map(parseEvent) }
	}

	if (properties !== undefined) {
		data.chat.properties = properties
	}

	return data
}

export const parseUpdateLastSeenTimestampData = data => {
	const parsed = {}
	parsed.chat_id = data.chat

	if (typeof data.timestamp === 'number') {
		parsed.timestamp = Math.floor(data.timestamp / 1000)
	}

	return parsed
}
