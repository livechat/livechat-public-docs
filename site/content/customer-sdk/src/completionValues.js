// @flow
import type { Success } from './types'

export const SUCCESS: Success = Object.freeze({ success: true })
