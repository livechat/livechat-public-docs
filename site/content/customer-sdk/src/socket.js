// @flow
import SockJS from 'sockjs-client'
import mitt from '@livechat/mitt'
import Backo from 'backo2'
import { buildQueryString } from '@livechat/url-utils'
import { debounce, generateRandomId } from '@livechat/data-utils'

import type { LicenseId, Request, SocketMessage } from './types'

type SocketEventCallback<Incoming> = {
	('connect', () => void): void,
	('message', (Incoming) => void): void,
	('disconnect', () => void): void,
}

type Socket<Outgoing, Incoming> = $ReadOnly<{|
	connect: () => void,
	disconnect: () => void,
	reconnect: () => void,
	emit: Outgoing => void,
	on: SocketEventCallback<Incoming>,
	off: SocketEventCallback<Incoming>,
|}>

export default {
	create(url: string, license: LicenseId): Socket<Request, SocketMessage> {
		const socketUrl = `${ url }?${ buildQueryString({
			license_id: license,
			bh: generateRandomId(),
		}) }`

		const emitter = mitt()
		const backoff = new Backo({
			min: 1000,
			max: 5000,
			jitter: 0.5,
		})

		let opened = false
		let connected = false
		let reconnectTimer = null
		let socket = null

		const ping = () => {
			// $FlowFixMe we know that that socket is defined when this gets called
			socket.send('{"action": "ping"}')
		}

		const schedulePing = debounce(30 * 1000, () => {
			if (!opened) {
				return
			}

			if (connected) {
				ping()
			}

			schedulePing()
		})

		const openHandler = () => {
			connected = true
			backoff.reset()
			emitter.emit('connect')
			schedulePing()
		}

		const closeHandler = () => {
			connected = false
			socket = null

			emitter.emit('disconnect')
			// eslint-disable-next-line no-use-before-define
			reconnect()
		}

		const messageHandler = ({ data }) => {
			const message = JSON.parse(data)

			if (message.action === 'ping') {
				return
			}

			emitter.emit('message', message)
		}

		const heartbeatListener = () => {
			ping()
			schedulePing()
		}

		const addEventListeners = instance => {
			instance.addEventListener('open', openHandler)
			instance.addEventListener('close', closeHandler)
			instance.addEventListener('message', messageHandler)
			instance.addEventListener('heartbeat', heartbeatListener)
		}

		const removeEventListeners = instance => {
			// $FlowFixMe
			instance.removeEventListener('open', openHandler)
			// $FlowFixMe
			instance.removeEventListener('close', closeHandler)
			// $FlowFixMe
			instance.removeEventListener('message', messageHandler)
			// $FlowFixMe
			instance.removeEventListener('heartbeat', heartbeatListener)
		}

		const close = () => {
			// $FlowFixMe we know that socket is defined at this point
			removeEventListeners(socket)
			connected = false
			// $FlowFixMe we know that socket is defined at this point
			socket.close()
			socket = null
		}

		const connect = () => {
			opened = true
			socket = new SockJS(socketUrl, undefined, { transports: ['websocket', 'xhr-polling'] })

			if (process.env.NODE_ENV !== 'production') {
				window.sockJS = socket
			}
			addEventListeners(socket)
		}

		const reconnect = (delay = backoff.duration()) => {
			clearTimeout(reconnectTimer)
			reconnectTimer = setTimeout(connect, delay)
		}

		return {
			connect: () => {
				if (opened) {
					throw new Error('Socket is already connected or opening.')
				}
				connect()
			},
			disconnect() {
				opened = false
				clearTimeout(reconnectTimer)

				if (socket === null) {
					return
				}

				close()
			},
			reconnect: delay => {
				if (!opened) {
					throw new Error('Socket is not opened.')
				}

				if (socket !== null) {
					close()
				}

				reconnect(delay)
			},
			emit: frame => {
				if (!connected) {
					throw new Error('Socket is not connected.')
				}

				// $FlowFixMe connected state means there is a defined socket
				socket.send(JSON.stringify(frame))
				schedulePing()
			},
			on: emitter.on,
			off: emitter.off,
		}
	},
}
