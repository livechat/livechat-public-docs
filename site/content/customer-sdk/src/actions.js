import { isObject } from '@livechat/data-utils'
import * as clientDataParsers from './clientDataParsers'

export const sendRequest = (action, data = null, meta = {}) => ({
	type: 'send_request',
	payload: {
		action,
		data,
	},
	meta,
})

export const sendEvent = (chat, data, meta) => {
	const event = clientDataParsers.parseEvent(data)
	const payload = {
		chat_id: chat,
		event,
	}
	if (isObject(meta)) {
		if (meta.attachToLastThread) {
			payload.attach_to_last_thread = true
		}
	}
	const action = sendRequest('send_event', payload)

	return action
}

export const setChatActive = (id, active) => ({ type: 'set_chat_active', payload: { id, active } })

export const storeUpdate = (updateAction, data) => ({
	type: 'store_update',
	payload: {
		updateAction,
		data,
	},
})
