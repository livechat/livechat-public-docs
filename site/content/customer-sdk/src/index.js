// @flow
import createAuth, { validateConfig } from '@livechat/customer-auth'
import mitt from '@livechat/mitt'
import $$observable from 'symbol-observable'
import createStore from './createStore'
import { isChatActive, isConnected, isLogging } from './reducer'
import chatHistory from './chatHistory'
import socketWrapper from './socket'
import socketListener from './socketListener'
import _sendRequestAction from './sendRequestAction'
import sendFile from './sendFile'
import * as actions from './actions'
import * as clientDataParsers from './clientDataParsers'
import { isObject, last, noop } from '@livechat/data-utils'
export { default as debug } from './debug'
import type {
	ChatId,
	CustomerPage,
	Env,
	EventId,
	EventOptionalProps,
	LicenseId,
	Timestamp,
	ThreadId,
	EventServerBase,
	Message,
	Event,
	Rating,
	Customer,
	ThreadSummary,
	Thread,
	ChatSummary,
	Chat,
	Postback,
	RNFile,
	RoutingScope,
	Success,
	SDKEvents,
	Pagination,
} from './types'

const CHATS_PAGINATION_MAX_LIMIT = 25
const THREADS_PAGINATION_MAX_LIMIT = 1000

export const init = (
	config: {
		autoConnect?: boolean,
		customer?: {
			name?: $ElementType<Customer, 'name'>,
			email?: $ElementType<Customer, 'email'>,
			fields?: $ElementType<Customer, 'fields'>,
		},
		license: LicenseId,
		group?: number,
		region?: string,
		clientId: string,
		page: CustomerPage,
		referrer?: string,
		uniqueGroups?: boolean,
	},
	env: Env = 'production',
) => {
	validateConfig(config)
	const { autoConnect = true, license, group, customer, region, page: initialPage, referrer, uniqueGroups } = config
	const envPart = (() => {
		if (license === 1520) {
			return '.staging'
		}
		if (env === 'production') {
			return ''
		}
		return `.${ env }`
	})()
	const regionPart = region ? `-${ region }` : ''
	const serverUrl = `https://api${ regionPart }${ envPart }.livechatinc.com/v3.0/customer`

	const emitter = mitt()
	const socket = socketWrapper.create(`${ serverUrl }/rtm/sjs`, license)
	const auth = createAuth(config, env)

	const store = createStore(
		{ license, group, customer, env, page: initialPage, referrer, uniqueGroups },
		emitter,
		socket,
		auth,
	)

	if (process.env.NODE_ENV !== 'production') {
		window.customerSocket = socket
	}

	socketListener(store, socket)

	const sendRequestAction = _sendRequestAction.bind(null, store)
	const startConnection = () => {
		store.dispatch({ type: 'start_connection' })
	}

	const api = Object.freeze({
		auth,
		closeThread(chat: ChatId): Promise<Success> {
			return sendRequestAction(actions.sendRequest('close_thread', { chat_id: chat }))
		},
		connect: startConnection,
		destroy() {
			store.dispatch({ type: 'destroy' })
		},
		disconnect() {
			store.dispatch({ type: 'stop_connection' })
		},
		getChatHistory(chat: ChatId): $AsyncIterator<Event[], Event[], void> {
			return chatHistory(store, api, chat)
		},
		getChatsSummary(pagination: Pagination = {}): Promise<{ chatsSummary: ChatSummary[], totalChats: number }> {
			const { offset = 0, limit = 10 } = pagination

			if (limit > CHATS_PAGINATION_MAX_LIMIT) {
				return Promise.reject(new Error(`Specified limit is too high (max ${ CHATS_PAGINATION_MAX_LIMIT }).`))
			}

			return sendRequestAction(actions.sendRequest('get_chats_summary', { offset, limit }))
		},
		getChatThreads(chat: ChatId, threads: ThreadId[]): Promise<Thread[]> {
			return sendRequestAction(
				actions.sendRequest('get_chat_threads', {
					chat_id: chat,
					thread_ids: threads,
				}),
			)
		},
		getChatThreadsSummary(
			chat: ChatId,
			pagination: Pagination = {},
		): Promise<{ threadsSummary: ThreadSummary[], totalThreads: number }> {
			const { offset = 0, limit = 25 } = pagination

			if (limit > THREADS_PAGINATION_MAX_LIMIT) {
				return Promise.reject(new Error(`Specified limit is too high (max ${ THREADS_PAGINATION_MAX_LIMIT }).`))
			}

			return sendRequestAction(
				actions.sendRequest('get_chat_threads_summary', {
					chat_id: chat,
					offset,
					limit,
				}),
			)
		},
		getPredictedAgent() {
			return sendRequestAction(actions.sendRequest('get_predicted_agent'))
		},
		getUrlDetails(url: string) {
			return sendRequestAction(actions.sendRequest('get_url_details', { url }))
		},
		on: (emitter.on: SDKEvents),
		once: (emitter.once: SDKEvents),
		off: (emitter.off: SDKEvents),
		rateChat(chat: ChatId, rating: $Diff<Rating, { ...EventServerBase, type: 'rating' }>): Promise<{ id: string }> {
			const event = { ...rating, type: 'rating' }

			// $FlowFixMe
			return Promise.all([api.sendEvent(chat, event).catch(() => ({})), api.getChatThreadsSummary(chat)])
				.then(([eventResponse, { threadsSummary }]) =>
					Promise.all([
						eventResponse,
						api.updateChatThreadProperties(
							chat,
							last(threadsSummary).id,
							clientDataParsers.parseRatingProperties(event),
						),
					]),
				)
				.then(([{ id }]) => ({ id }))
		},
		sendEvent<T: Event>(chat: ChatId, event: $Diff<T, EventServerBase>, meta): Promise<T> {
			return sendRequestAction(actions.sendEvent(chat, event, meta))
		},
		sendFile(
			chat: ChatId,
			data: {
				...EventOptionalProps,
				file: Blob | RNFile,
			},
			spec: {
				onProgress: number => mixed,
			},
		) {
			return sendFile({ auth, chat, license, serverUrl, data, spec })
		},
		sendPostback(chat: ChatId, thread: ThreadId, event: EventId, { id, toggled = true }: Postback): Promise<Postback> {
			return sendRequestAction(
				actions.sendRequest('send_rich_message_postback', {
					chat_id: chat,
					event_id: event,
					thread_id: thread,
					postback: {
						id,
						toggled,
					},
				}),
			)
		},
		setSneakPeek: (chat: ChatId, text: string) => {
			if (!isChatActive(store.getState(), chat)) {
				return
			}
			sendRequestAction(
				actions.sendRequest('send_sneak_peek', {
					chat_id: chat,
					sneak_peek_text: text,
				}),
			)
		},
		startChat(
			data: {
				scope?: RoutingScope,
				events?: Array<$Diff<Event, EventServerBase>>,
				// TODO: uncomment when configuration API is done
				// properties?: Properties,
			} = {},
		): Promise<Chat> {
			return sendRequestAction(actions.sendRequest('start_chat', clientDataParsers.parseStartChatData(data)))
		},
		updateChatProperties(chat: ChatId, properties: Properties): Promise<Success> {
			return sendRequestAction(
				actions.sendRequest('update_chat_properties', {
					chat_id: chat,
					properties,
				}),
			)
		},
		updateChatThreadProperties(chat: ChatId, thread: ThreadId, properties: Properties): Promise<Success> {
			return sendRequestAction(
				actions.sendRequest('update_chat_thread_properties', {
					chat_id: chat,
					thread_id: thread,
					properties,
				}),
			)
		},
		updateCustomer(properties: {
			name?: $ElementType<Customer, 'name'>,
			email?: $ElementType<Customer, 'email'>,
			fields?: $ElementType<Customer, 'fields'>,
		}): void {
			store.dispatch({ type: 'update_customer', payload: properties })

			const state = store.getState()

			if (isConnected(state)) {
				sendRequestAction(actions.sendRequest('update_customer', properties)).catch(noop)
			} else if (isLogging(state)) {
				store.dispatch(actions.storeUpdate('update_customer', properties))
			}
		},
		updateCustomerPage(page: CustomerPage): void {
			store.dispatch({ type: 'update_customer_page', payload: page })

			const state = store.getState()

			if (isConnected(state)) {
				sendRequestAction(actions.sendRequest('update_customer_page', page)).catch(noop)
			} else if (isLogging(state)) {
				store.dispatch(actions.storeUpdate('update_customer_page', page))
			}
		},
		updateEventProperties(chat: ChatId, thread: ThreadId, event: EventId, properties: Properties): Promise<Success> {
			return sendRequestAction(
				actions.sendRequest('update_event_properties', {
					chat_id: chat,
					event_id: event,
					thread_id: thread,
					properties,
				}),
			)
		},
		updateLastSeenTimestamp(chat: ChatId, timestamp?: Timestamp): Promise<{ timestamp: Timestamp }> {
			return sendRequestAction(
				actions.sendRequest(
					'update_last_seen_timestamp',
					clientDataParsers.parseUpdateLastSeenTimestampData({ chat, timestamp }),
				),
			)
		},
		[$$observable]() {
			return {
				subscribe(observer) {
					if (!isObject(observer)) {
						throw new TypeError('Expected the observer to be an object.')
					}
					const subscriber = (type, event) => {
						if (typeof observer.next !== 'function') {
							return
						}
						observer.next([type, event])
					}
					emitter.on('*', subscriber)
					return () => emitter.off('*', subscriber)
				},
			}
		},
	})

	if (autoConnect) {
		startConnection()
	}

	return api
}
