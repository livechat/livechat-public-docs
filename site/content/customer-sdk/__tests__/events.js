jest.mock('@livechat/customer-auth', () => ({
	__esModule: true,
	validateConfig: ({ license, clientId }) => {
		if (typeof license !== 'number' || typeof clientId !== 'string') {
			throw new Error('You need to pass valid configuration object: { license, clientId }.')
		}
	},
	default: license => ({
		getToken() {
			return Promise.resolve({
				accessToken: String(Math.random()),
				entityId: 'test_customer_id',
				expiresIn: 28800000,
				tokenType: 'Bearer',
				creationDate: Date.now(),
				license,
			})
		},
	}),
}))

jest.mock('../src/socket')

import { last } from '@livechat/data-utils'
import * as CustomerSDK from '../src'
import * as socketModule from '../src/socket'

const KILOBYTE = 1024
const MEGABYTE = 1024 * KILOBYTE

const initSDK = () => CustomerSDK.init({ license: 123, clientId: 'test_client_id' })

describe('CustomerSDK - events sending', () => {
	const originalNow = Date.now
	beforeAll(() => {
		const now = Date.now()
		let counter = 0
		Date.now = () => now + ++counter * 60 * 1000
		Date.now.getTime = () => now + counter * 60 * 1000
	})
	beforeEach(() => {
		socketModule.server.clearHistory()
	})
	afterAll(() => {
		Date.now = originalNow
	})

	test('should send event to active chat', done => {
		socketModule.server.addChats(['2_threads_active'])
		const sdk = initSDK()
		sdk.on('connected', data => {
			const chat = data.chatsSummary[0].id
			const lastThread = data.chatsSummary[0].lastEvent.thread

			sdk
				.sendEvent(chat, {
					type: 'message',
					text: 'test msg',
				})
				.then(message => {
					expect(message).toEqual({
						type: 'message',
						text: 'test msg',
						id: 'sent_event_1',
						author: 'test_customer_id',
						timestamp: Math.floor(Date.now.getTime() / 1000) * 1000,
						thread: lastThread,
						properties: {},
					})
					done()
				})
		})
	})

	test('should send message to inactive chat', done => {
		const actual = []
		const expected = [
			{ fields: {}, id: 'test_customer_id', type: 'customer' },
			{
				avatar: 'https://cdn.livechat-static.com/avatars/1.png',
				id: 'fixture_agent_1',
				name: 'John Doe',
				type: 'agent',
			},
			{
				active: true,
				chat: 'fixture_chat_2',
				id: 'created_thread_1',
				order: 3,
				totalEvents: 1,
				properties: {},
			},
			{
				type: 'message',
				text: 'test msg',
				id: 'sent_event_1',
				author: 'test_customer_id',
				timestamp: null,
				thread: 'created_thread_1',
				properties: {},
			},
		]
		socketModule.server.addChats(['2_threads_inactive'])
		const sdk = initSDK()
		sdk.on('connected', data => {
			const chat = data.chatsSummary[0].id

			sdk.on('user_data', threadSummary => {
				actual.push(threadSummary)
			})
			sdk.on('thread_summary', threadSummary => {
				actual.push(threadSummary)
			})

			sdk
				.sendEvent(chat, {
					type: 'message',
					text: 'test msg',
				})
				.then(message => {
					actual.push(message)
					// ugly as hell, need to figure out timestamp assertions
					last(expected).timestamp = Math.floor(Date.now.getTime() / 1000) * 1000
					expect(actual).toEqual(expected)
					done()
				})
		})
	})

	test('should send event with customId', done => {
		socketModule.server.addChats(['2_threads_active'])
		const sdk = initSDK()
		sdk.on('connected', data => {
			const chat = data.chatsSummary[0].id
			const lastThread = data.chatsSummary[0].lastEvent.thread

			sdk
				.sendEvent(chat, {
					type: 'message',
					text: 'test msg',
					customId: 'test_custom_id',
				})
				.then(message => {
					expect(message).toEqual({
						type: 'message',
						text: 'test msg',
						id: 'sent_event_1',
						author: 'test_customer_id',
						timestamp: Math.floor(Date.now.getTime() / 1000) * 1000,
						thread: lastThread,
						customId: 'test_custom_id',
						properties: {},
					})
					done()
				})
		})
	})

	test('should fail when sending message without text', done => {
		socketModule.server.addChats(['2_threads_active'])
		const sdk = initSDK()
		sdk.on('connected', data => {
			const chat = data.chatsSummary[0].id

			sdk.sendEvent(chat, { type: 'message' }).catch(err => {
				expect(err.code).toEqual('REQUEST_FAILED')
				done()
			})
		})
	})

	test('should send unknown event types in passthrough manner', done => {
		socketModule.server.addChats(['2_threads_active'])
		const sdk = initSDK()
		sdk.on('connected', data => {
			const chat = data.chatsSummary[0].id
			const lastThread = data.chatsSummary[0].lastEvent.thread

			sdk
				.sendEvent(chat, {
					type: 'unknown_type',
					customId: 'test_custom_id',
					unknown_prop: 1,
					unknown_prop2: 'test',
					unknown_prop3: {
						whatever: true,
					},
				})
				.then(message => {
					expect(message).toEqual({
						type: 'unknown_type',
						id: 'sent_event_1',
						author: 'test_customer_id',
						timestamp: Math.floor(Date.now.getTime() / 1000) * 1000,
						thread: lastThread,
						customId: 'test_custom_id',
						unknown_prop: 1,
						unknown_prop2: 'test',
						unknown_prop3: {
							whatever: true,
						},
						properties: {},
					})
					done()
				})
		})
	})

	test.skip('should send file and receive the event as push', done => {
		done()
	})

	test.skip('should send file with max size', done => {
		done()
	})

	test('should not send too big file', done => {
		socketModule.server.addChats(['2_threads_active'])
		const sdk = initSDK()
		sdk.on('connected', data => {
			const chat = data.chatsSummary[0].id
			expect(sdk.sendFile(chat, { file: { size: 10 * MEGABYTE + 1 } }))
				.rejects.toThrowErrorMatchingSnapshot()
				.then(done)
		})
	})

	test.skip('should send file and be notified about the progress', done => {
		done()
	})

	test.skip('should be able to cancel file upload', done => {
		done()
	})
})
