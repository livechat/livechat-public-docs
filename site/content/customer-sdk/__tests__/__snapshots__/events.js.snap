// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerSDK - events sending should not send too big file 1`] = `"The file is too big (max size is 10.00 MB)."`;
